src/net_processing.cpp

```diff
@@ -4701,6 +4701,21 @@ void PeerManagerImpl::ProcessMessage(CNode& pfrom, const std::string& msg_type,
         const uint256& hash = peer->m_wtxid_relay ? wtxid : txid;
         AddKnownTx(*peer, hash);
 
+        if (auto num_broadcasted = m_tx_for_private_broadcast.Remove(ptx)) {
+            LogPrintLevel(BCLog::PRIVATE_BROADCAST,
+                          BCLog::Level::Info,
+                          "Received our privately broadcast transaction (txid=%s) from the "
+                          "network from peer=%d%s; stopping private broadcast attempts\n",
+                          txid.ToString(),
+                          pfrom.GetId(),
+                          fLogIPs ? strprintf(", peeraddr=%s", pfrom.addr.ToStringAddrPort()) : "");
+            if (NUM_PRIVATE_BROADCAST_PER_TX > num_broadcasted.value()) {
+                // Not all of the initial NUM_PRIVATE_BROADCAST_PER_TX connections were needed.
+                // Tell CConnman it does not need to start the remaining ones.
+                m_connman.m_private_broadcast.NumToOpenSub(NUM_PRIVATE_BROADCAST_PER_TX - num_broadcasted.value());
```

Comment:
3f5a4187a9e5d2bbb65c06c1799155fd7dd0d132

Feels like this should be part of `Remove()`, which I think highlights some possible confusion between the two `class PrivateBroadcast` you have. One in ConnMan and one in its own unit, used in PeerManager.

I dunno if there's a cleaner approach available, to ensure that the two objects don't get out of sync? Like instead of `m_num_to_open` in ConnMan, could you refer to the size of the `ByNodeId` map?

Reply:
There is some layer separation between `net`/`CConnman` and `net_processing`/`PeerManager` which I am trying not to blur too much. The lower `net` level only knows about connections and things like "I need to open N connections of type X". It does not know anything about transactions, or P2P messages which are handled by the `net_processing` layer.

The constant `NUM_PRIVATE_BROADCAST_PER_TX` (_send a transaction to this many peers_) belongs to the `net_processing` layer. Calling this `NumToOpenSub()` is only done if `Remove()` succeeds and `NUM_PRIVATE_BROADCAST_PER_TX > num_broadcasted.value()` holds. So, no way to do it in `CConnman` because that would mean accessing `NUM_PRIVATE_BROADCAST_PER_TX` from `CConnman`.

> could you refer to the size of the `ByNodeId` map?

That is a "transaction id mapped by node id". Couldn't deal with transactions from within `CConnman`.